platform/drm/cog-drm-modeset-renderer.c

```diff
@@ -0,0 +1,694 @@
+    wl_resource_add_destroy_listener(buffer_resource, &buffer->destroy_listener);
+    wl_resource_set_user_data(buffer_resource, self);
+
+    buffer->fb_id = fb_id;
+    buffer->bo = bo;
+    buffer->buffer_resource = buffer_resource;
+
+    return buffer;
+}
+
+static struct buffer_object *
+drm_create_buffer_for_shm_buffer(CogDrmModesetRenderer *self,
+                                 struct wl_resource    *buffer_resource,
+                                 struct wl_shm_buffer  *shm_buffer)
+{
+    uint32_t format = wl_shm_buffer_get_format(shm_buffer);
+    if (format != WL_SHM_FORMAT_ARGB8888 && format != WL_SHM_FORMAT_XRGB8888) {
+        g_warning("failed to handle non-32-bit ARGB/XRGB format");
+        return NULL;
+    }
+
+    int32_t width = wl_shm_buffer_get_width(shm_buffer);
+    int32_t height = wl_shm_buffer_get_height(shm_buffer);
+
+    // TODO: don't ignore the alpha channel in case of ARGB8888 SHM data
+    uint32_t       gbm_format = GBM_FORMAT_XRGB8888;
+    struct gbm_bo *bo = gbm_bo_create(self->gbm_dev, width, height, gbm_format, GBM_BO_USE_SCANOUT | GBM_BO_USE_WRITE);
+    if (!bo) {
+        g_warning("failed to create a gbm_bo object");
+        return NULL;
+    }
+
+    uint32_t in_handles[4] = {
+        0,
+    };
+    uint32_t in_strides[4] = {
+        0,
+    };
+    uint32_t in_offsets[4] = {
+        0,
+    };
+    in_handles[0] = gbm_bo_get_handle(bo).u32;
+    in_strides[0] = gbm_bo_get_stride(bo);
+
+    uint32_t fb_id = 0;
+    int ret = drmModeAddFB2(get_drm_fd(self), width, height, gbm_format, in_handles, in_strides, in_offsets, &fb_id, 0);
+    if (ret) {
+        gbm_bo_destroy(bo);
+        g_warning("failed to create framebuffer: %s", g_strerror(errno));
+        return NULL;
+    }
+
+    struct buffer_object *buffer = g_new0(struct buffer_object, 1);
+    wl_list_insert(&self->buffer_list, &buffer->link);
+    buffer->destroy_listener.notify = destroy_buffer_notify;
+    wl_resource_add_destroy_listener(buffer_resource, &buffer->destroy_listener);
+    wl_resource_set_user_data(buffer_resource, self);
+
+    buffer->fb_id = fb_id;
+    buffer->bo = bo;
+    buffer->buffer_resource = buffer_resource;
+
+    return buffer;
+}
+
+static void
+drm_copy_shm_buffer_into_bo(struct wl_shm_buffer *shm_buffer, struct gbm_bo *bo)
+{
+    int32_t width = wl_shm_buffer_get_width(shm_buffer);
+    int32_t height = wl_shm_buffer_get_height(shm_buffer);
+    int32_t stride = wl_shm_buffer_get_stride(shm_buffer);
+
+    uint32_t bo_stride = 0;
+    void    *map_data = NULL;
+    gbm_bo_map(bo, 0, 0, width, height, GBM_BO_TRANSFER_WRITE, &bo_stride, &map_data);
+    if (!map_data)
+        return;
+
+    wl_shm_buffer_begin_access(shm_buffer);
+
+    uint8_t *src = wl_shm_buffer_get_data(shm_buffer);
+    uint8_t *dst = map_data;
+
+    uint32_t bo_width = gbm_bo_get_width(bo);
+    uint32_t bo_height = gbm_bo_get_height(bo);
+    if (!(width == bo_width && height == bo_height && stride == bo_stride)) {
```

Comment:
If the dimensions match, including the stride, that means the allocated size for both buffers is the same, and that the disposition of pixels inside them match because pixel format is the same for both. That is the reason why there are no bounds check in the `else` part.

The BO is created with the same dimensions and pixel format as SHM buffers the first time it is needed, so I would expect that at least width and height always match, given that `wpe_view_backend_dispatch_set_size()` is never being called in the DRM/KMS platform plug-in.

OTOH, when the stride does not match, I do wonder if it's faster to let the compiler derive the memory copy with a double loop, or whether it would be better to do a row-by-row copy, or maybe even offload to Pixman, which is likely to have hand-optimized assembler pixel copy routines for most common platforms 🤔 

FWIW, this was existing code that got moved around, so I would rather leave it as-is for now. Originally this was written by @zdobersek, so if he wants to comment here about possible improvements, I will be happy to do a follow-up patch.